### DISCUSSION TITLE
함수형 컴포넌트와 클래스형 컴포넌트를 비교해 설명하세요.

### DISCUSSION THREAD

**Discussion by bona373737** [wanted-9th-3team/frontend-study#32] [프론트 기술 질문 공간]
함수형 컴포넌트와 클래스형 컴포넌트를 비교해 설명하세요.

**Comment by michoball** [comment]
과거 리액트의 컴포넌트는 클래스형 컴포넌트 였습니다.  state나 life cycle에 관련된 기능들을 사용할 수 있고, 임의 메서드를 생성할 수 있지만, 컴포넌트 내 함수에 this를 사용해야하고 render()로 JSX를 render해야하고 메모리가 좀 더 많이 소모된다는 등의 특징도 가지고 있습니다. 

비교적 최근에 나온 함수형 컴포넌트는 life cycle 과 관련된 기능이 있지는 않지만 훅이 도입되면서 많은 부분에서 클래스 컴포넌트 보다 장점을 가지고 있고, 리액트 공식문서에서도 hook을 사용하기를 권장하고 있습니다.

**Comment by WooGie911** [comment]
React 컴포넌트를 만들때 클래스형 컴포넌트, 함수형 컴포넌트 2가지 방식이 있습니다.

과거에는 클래스형 컴포넌트를 많이 사용했지만 2019년 v16.8 부터 함수형 컴포넌트에 리액트 훅(hook)을 지원하며

현재는 공식 문서에서도 함수형 컴포넌트와 훅(hook)을 함께 사용할 것을 권장하고 있습니다. 

클래스형 컴포넌트와 함수형 컴포넌트의 역할은 동일합니다.

차이점으로는 5가지가 있습니다.

1. 선언방식

클래스형 컴포넌트
- class 키워드가 필요하다.
- Component로 상속을 받아야 한다.
- render() 메소드가 반드시 필요하다.
- state, lifeCycle 관련 기능 사용이 가능하다.
- 함수형보다 메모리 자원을 더 사용한다.
- 임의 메소드를 정의할 수 있다.

함수형 컴포넌트
- state, lifeCycle 관련 기능사용 불가능하다. (Hook을 통해 해결)
- 클래스형보다 메모지 자원을 덜 사용한다.
- 컴포넌트 선언이 편하다.


2. state

클래스형 컴포넌트
- constructor 안에서 this.state 초기 값 설정 가능하다.
- counstructor 없이도 바로 state 초기값을 설정 가능하다.
- this.setState() 를 통해 state값을 변경 할 수 있다.
- 클래스형의 state는 객체형식이다.

함수형 컴포넌트
- useState 함수로 state를 사용한다.
- useState 함수를 호출하면 배열이 반환되는데 첫 번째 원소는 현재 상태, 두번째 원소는 상태를 바꿔주는 함수이다.


3. props

: 클래스형 컴포넌트는 this.props로 통해 값을 불러올 수 있으나 함수형 컴포넌트는 props를 불러올 필요 없이 바로 호출 할 수 있다.


4. 이벤트 핸들링

클래스형 컴포넌트
- 함수 선언시 화살표 함수로 바로 선언 가능하다.
- 요소에 적용할때 this.를 붙여 줘야 한다.

함수형 컴포넌트
- const + 함수 형태로 선언 해야 한다.
- 요소에 적용할때 this가 필요없다.


5. 생명주기 ( LifeCycle )

: 클래스형 컴포넌트에서는 LifeCycle API를 사용하며, 함수형 컴포넌트에서는 Hook을 사용한다.

**Comment by skyhanull** [comment]
클레스형 컴포넌트는 life cycle이 존재하고 class로 시작을 합니다. 또한 render()를 사용하여 jsx를 반환하고 props를 조회할 때 this 키워드 사용합니다
함수형 컴포넌트는 react hook을 사용하는 것으로 life cycle은 없지만 useEffect로 대신할 수 있으며 클레스형의 기능을 대부분 사용 할 수 있다고 합니다.
